fix: delete subtasks from task that crashed while processing

italo1aguiar-cacib opened the pull request:
# Motivation

When a pod crashes while processing a task near completion, other pods might try to handle this crash when they get the task message from the queue. When two pods try to acquire this task at the same time and handle this task simultaneously, they might end up cancelling a task while the other is still checking created subtasks status. Since Armonik heuristics consider that a task must have all subtasks on status creating while its parent task is no completed, in a concurrency case where a task is cancelled during the check of this heuristic, the parent task is considered completed by the other pods and has its status updated on mongo, making it blocked. This causes non-reexecution of this and blocks the workflow for all dependent tasks.

# Description

This PR fixes the race condition by introducing a new PurgeTasksAndAbortResults method that deletes subtasks instead of just cancelling them when a task crashes during processing.

# Testing

Updated unit tasks and global stress tests at CACIB

# Impact

Subtasks from crashed tasks are now deleted instead of marked as cancelled, preventing the race condition that could block the workflow.

# Additional Information

The recursion kept over tasks dependent indrectly might be removed after discussion, it might not be necessary. Results can also be deleted. The intent of this commit is to fix the bug that was locking sessions, keeping the behaviour close as possible.

# Checklist

- [x] My code adheres to the coding and style guidelines of the project.
- [x] I have performed a self-review of my code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] I have thoroughly tested my modifications and added tests when necessary.
- [x] Tests pass locally and in the CI.
- [ ] I have assessed the performance impact of my modifications.

idhir-Cacib-armoniK commented on Common/src/Storage/ResultLifeCycleHelper.cs:
```diff
@@ -185,6 +185,140 @@ await AbortTasksAndResults(taskTable,
+    if (!taskIds.Any() && !resultIds.Any())
+    {
+      return;
+    }
```
As it is array it is pereferable to use 

taskIds.Length != 0 && resultIds.Length != 0 

maybe a log in debug to say no purge to do :wink:

italo1aguiar-cacib replied:
since it is a collection, so we must use Count, done

italo1aguiar-cacib replied:
btw, it should be the inverse (==)

idhir-Cacib-armoniK commented on Common/src/Storage/ResultLifeCycleHelper.cs:
```diff
@@ -185,6 +185,140 @@ await AbortTasksAndResults(taskTable,
+                         .Concat(resultIds)
+                         .ToHashSet();
+      }
+    }
```
please create a methods named TryGetResultsFromTaskIds() where you put the code from line 232 to 271

italo1aguiar-cacib replied:
done

idhir-Cacib-armoniK commented on Common/src/Storage/ResultLifeCycleHelper.cs:
```diff
@@ -185,6 +185,140 @@ await AbortTasksAndResults(taskTable,
+
+      // Find all metadata about the tasks that must be aborted
+      var tasks = await taskTable.FindTasksAsync(task => taskIds.Contains(task.TaskId) &&
+                                                         (task.Status == TaskStatus.Creating || task.Status == TaskStatus.Pending || task.Status == TaskStatus.Paused ||
```
maybe create an HashSet with all the TaskStatus in the status needed to be retrieved like this 

```cs
var taskStatusToSeek = new HashSet<TaskStatus> 
{
TaskStatus.Creating, ... TaskStatus.Timeout
}

// then in the filter FindTasksAsync do 

taskStatusToSeek.Contains(task.Status)
```

italo1aguiar-cacib replied:
done

idhir-Cacib-armoniK commented on Common/src/Storage/ResultLifeCycleHelper.cs:
```diff
@@ -185,6 +185,140 @@ await AbortTasksAndResults(taskTable,
+
+    // Early exit if no result has been aborted
+    if (count == 0)
+    {
```
Debug Log welcomed here

italo1aguiar-cacib replied:
done

idhir-Cacib-armoniK commented on Common/src/Storage/ResultLifeCycleHelper.cs:
```diff
@@ -185,6 +185,140 @@ await AbortTasksAndResults(taskTable,
+                                    resultIds,
+                                    reason);
+
+    var count = await resultTable.UpdateManyResults(result => resultIds.Contains(result.ResultId) &&
```
maybe I'm unfarmiliar with mongo, but here we will scan all the documents no ? 
Is there a better way here to update maybe with filter ?

italo1aguiar-cacib replied:
Hashed index on OwnerTaskId so it will run over a subset of results

idhir-Cacib-armoniK commented on Common/src/Storage/ResultLifeCycleHelper.cs:
```diff
@@ -185,6 +185,140 @@ await AbortTasksAndResults(taskTable,
+
+    // Find all the tasks that depend on the aborted results
+    var nextTasks = new HashSet<string>();
+    await foreach (var dependentTasks in resultTable.GetResults(result => resultIds.Contains(result.ResultId) && result.Status == ResultStatus.Aborted &&
```
same comment regarding the filter of data ?

italo1aguiar-cacib replied:
same comment

idhir-Cacib-armoniK commented on Common/src/Storage/TaskLifeCycleHelper.cs:
```diff
@@ -1035,7 +1034,7 @@ await ResultLifeCycleHelper.AbortTasksAndResults(taskTable,
                                         pushQueueStorage,
                                         taskData,
                                         sessionData,
-                                        subtasks.ViewSelect(td => td.TaskId),
```
why remove this ?

italo1aguiar-cacib replied:
because on PurgeTasksAndAbortResults we already operated over these taskIds, we don't need to touch them anymore.

idhir-Cacib-armoniK reviewed (commented):
Got some questions and also some refacto to do for better management please

PierrePicard20 committed:
fix: CreateResults must preserve order of results

PierrePicard20 committed:
fix: CreateResults must preserve order of results

PierrePicard20 committed:
automated source code formating

PierrePicard20 committed:
fix: CreateResults must preserve results order (#1128)

# Motivation

CreateResults must preserve results order.

# Description

In CreateResults avoid the use of GetResults() which reshuffle the
results collection.

# Testing

Unit test added to enforce the fix.

# Impact

Results creation

# Additional Information

None

# Checklist

- [x] My code adheres to the coding and style guidelines of the project.
- [x] I have performed a self-review of my code.
- [ ] I have commented my code, particularly in hard-to-understand
areas.
- [ ] I have made corresponding changes to the documentation.
- [x] I have thoroughly tested my modifications and added tests when
necessary.
- [x] Tests pass locally and in the CI.
- [x] I have assessed the performance impact of my modifications.

lemaitre-aneo committed:
Add windows compatible image build

lemaitre-aneo committed:
Print Logs in CI for forks

PierrePicard20 committed:
fix: ImportResultsData must preserve order of results

PierrePicard20 committed:
Merge branch 'main' into ppi/blobOrder

lemaitre-aneo committed:
Build images in test jobs on fork repos

PierrePicard20 committed:
fix: ImportResultsData must preserve order of results

lemaitre-aneo committed:
Lower hardware constraints on RabbitMQ

lemaitre-aneo committed:
ci: Make pipeline compatible with forked repos (#1131)

# Motivation

Pipelines from forks fail because the pipeline cannot upload images.

# Description

Adjust the pipeline jobs if the repo is a fork:
- Skip the jobs for image building and image scanning
- Build the images inside the test jobs
- Skip logs uploading
- Show logs within the pipeline

In addition, dockerfiles have been updated to be built on Windows
without buildx.

# Testing

The new pipeline has been tested with the dummy PR
https://github.com/aneoconsulting/ArmoniK.Core/pull/1130

# Impact

For base repo, the pipeline remains the same. For instance, the images
are still built in dedicated jobs, and pulled from test jobs.
For forked repos, the pipeline is expected to be slower because it needs
to build images much more.

# Checklist

- [x] My code adheres to the coding and style guidelines of the project.
- [x] I have performed a self-review of my code.
- [x] I have commented my code, particularly in hard-to-understand
areas.
- [ ] I have made corresponding changes to the documentation.
- [x] I have thoroughly tested my modifications and added tests when
necessary.
- [x] Tests pass locally and in the CI.
- [x] I have assessed the performance impact of my modifications.

PierrePicard20 committed:
Merge branch 'main' into ppi/blobOrder

PierrePicard20 committed:
automated code formating

PierrePicard20 committed:
fix: ImportResultsData must preserve order of results

aneojgurhem committed:
fix: ImportResultsData must preserve order of results (#1133)

# Motivation

ImportResultsData must preserve order of results

# Description

In ImportResultsData avoid the use of GetResults() which reshuffle the
results collection.

# Testing

Unit test added to enforce the fix.

# Impact

Result import.

# Additional Information

None.

# Checklist

- [x] My code adheres to the coding and style guidelines of the project.
- [x] I have performed a self-review of my code.
- [ ] I have commented my code, particularly in hard-to-understand
areas.
- [ ] I have made corresponding changes to the documentation.
- [x] I have thoroughly tested my modifications and added tests when
necessary.
- [x] Tests pass locally and in the CI.
- [x] I have assessed the performance impact of my modifications.

aneojgurhem committed:
fix: ignore url not found when searching dockerhub for images to remove

aneojgurhem committed:
fix: ignore url not found when searching dockerhub for images to remove (#1135)

# Motivation

Job that removes our old docker images once a week fails on 404 when url
returned by DockerHub is invalid.

# Description

Modify script to treat 404 errors as warning instead of stopping the
script.

# Testing

- It was tested locally and the error was reproduced. Now script does
not fail and print a warning message when encountering 404 error.
- Pipeline was run manually on this branch with command `gh workflow run
.github/workflows/remove-artefacts.yml --ref jg/fix_remove` but the
error did not happen.

# Impact

- Pipeline working properly.

italo1aguiar-cacib committed:
fix: delete subtasks from task that crashed while processing

idhir-Cacib-armoniK commented on Adaptors/Embed/src/ObjectStorage.cs:
```diff
@@ -88,5 +89,6 @@ public Task TryDeleteAsync(IEnumerable<byte[]> ids,
     => Task.FromResult<IDictionary<byte[], long?>>(ids.ToDictionary(id => id,
-                                                                    id => (long?)id.LongLength));
+                                                                    id => (long?)id.LongLength,
+                                                                    new ByteArrayComparer()));
```
Why not having it as an external variable ? or Within a property ?

idhir-Cacib-armoniK commented on Adaptors/LocalStorage/src/ObjectStorage.cs:
```diff
@@ -206,7 +207,8 @@ await TryDeleteAsync(key,
                                                                     id => GetFileSize(Path.Combine(path_,
-                                                                                                   Encoding.UTF8.GetString(id)))));
+                                                                                                   Encoding.UTF8.GetString(id))),
+                                                                    new ByteArrayComparer()));
```
Same here ?

idhir-Cacib-armoniK commented on Adaptors/Memory/src/ObjectStorage.cs:
```diff
@@ -110,7 +111,8 @@ public Task TryDeleteAsync(IEnumerable<byte[]> ids,
     => Task.FromResult<IDictionary<byte[], long?>>(ids.ToDictionary(id => id,
-                                                                    Exists));
+                                                                    Exists,
+                                                                    new ByteArrayComparer()));
```
Same here ...

idhir-Cacib-armoniK commented on Adaptors/Redis/src/ObjectStorage.cs:
```diff
@@ -162,7 +163,9 @@ public async Task TryDeleteAsync(IEnumerable<byte[]> ids,
                 .ToDictionaryAsync(tuple => tuple.id,
-                                   tuple => tuple.Item2)
+                                   tuple => tuple.Item2,
+                                   new ByteArrayComparer(),
```
hmm same here

idhir-Cacib-armoniK commented on Adaptors/S3/src/ObjectStorage.cs:
```diff
@@ -224,7 +225,9 @@ public async Task TryDeleteAsync(IEnumerable<byte[]> ids,
                 .ToDictionaryAsync(tuple => tuple.id,
-                                   tuple => tuple.Item2)
+                                   tuple => tuple.Item2,
+                                   new ByteArrayComparer(),
```
same here ...

idhir-Cacib-armoniK commented on Common/tests/Helpers/SimpleObjectStorage.cs:
```diff
@@ -44,7 +45,8 @@ public Task TryDeleteAsync(IEnumerable<byte[]> ids,
     => Task.FromResult<IDictionary<byte[], long?>>(ids.ToDictionary(id => id,
-                                                                    _ => (long?)42));
+                                                                    _ => (long?)42,
+                                                                    new ByteArrayComparer()));
```
Same here...
